mypy/maptype.py

```diff
@@ -1,8 +1,61 @@
 from __future__ import annotations
 
-from mypy.expandtype import expand_type_by_instance
+from typing import cast
+
+from mypy.expandtype import expand_type, expand_type_by_instance
 from mypy.nodes import TypeInfo
-from mypy.types import AnyType, Instance, TupleType, TypeOfAny, has_type_vars
+from mypy.types import AnyType, Instance, TupleType, Type, TypeOfAny, has_type_vars
+
+
+def as_type(typ: Type, direction: int, target: Type) -> Type | None:
+    """Attempts to solve type variables in `target` so that `typ` is a subtype/supertype of
+    the resulting type.
+
+    Args:
+        typ: The type to map from.
+        direction: One of SUBTYPE_OF or SUPERTYPE_OF
+        target: The target instance type to map to.
+
+    Returns:
+        None: if the mapping is not possible.
+        Type: the mapped type if the mapping is possible.
+
+    Examples:
+        (list[int], Iterable[T]) -> Iterable[int]
+        (list[list[int]], Iterable[list[T]]) -> Iterable[list[int]]
+        (dict[str, int], Mapping[K, int]) -> Mapping[str, int]
+        (list[int], Mapping[K, V]) -> None
+    """
+    from mypy.subtypes import is_subtype
+    from mypy.typeops import get_all_type_vars
```

Comment:
had to put these there due to circular import issues

Reply:
This function should _definitely_ not be in this module (ideally we shouldn't even have the existing typeops import below, but that is a separate story).

Also the scope of this function is misleadingly broad. It should probably accept a `TypeInfo` as `target` (which probably must be a protocol), and then use `fill_typevars(...)` as a constraint inference target.

Finally, it is worth doing some performance measurements, we don't want any visible slow-down for something that is only needed for rare edge cases.